drgn_tools/locking.py

```diff
@@ -85,3 +229,238 @@ def for_each_mutex_waiter(prog: Program, mutex: Object) -> Iterable[Object]:
+
+# Linux kernel commit 617f3ef95177840c77f59c2aec1029d27d5547d6 ("locking/rwsem:
+# Remove reader optimistic spinning") (in v5.11) removed optimistic spinning for
+# readers and hence left one bit to check for spinnable
+_RWSEM_NONSPINNABLE = 1 << 1
+
+
+def rwsem_has_spinner(rwsem: Object) -> bool:
+    """
+    Check if rwsem has optimistic spinners or not
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: True if rwsem has optimistic spinners, False otherwise.
+    """
+    return osq_is_locked(rwsem.osq.address_of_())
+
+
+def for_each_rwsem_waiter_entity(rwsem: Object) -> Iterable[Object]:
+    """
+    Find rwsem_waiter(s) for given rwsem
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: Iterator of ``struct rwsem_waiter``
+    """
+
+    for waiter in list_for_each_entry(
+        "struct rwsem_waiter", rwsem.wait_list.address_of_(), "list"
+    ):
+        yield waiter
+
+
+def get_rwsem_owner(rwsem: Object) -> Object:
+    """
+    Find owner of  given rwsem
+
+    :param rwsem: ``struct rw_semaphore *``
+    :returns: ``struct task_struct *`` if owner can be found, NULL otherwise
+    """
+    prog = rwsem.prog_
+    if not rwsem.count.counter.value_():
+        print("rwsem is free.")
+        return NULL(prog, "struct task_struct *")
+
+    if is_rwsem_writer_owned(rwsem):
+        if (rwsem.owner.type_.type_name() != "atomic_long_t") and (
+            rwsem.owner.value_() & _RWSEM_ANONYMOUSLY_OWNED
+        ):
+            print("rwsem is owned by anonymous writer")
+            return NULL(prog, "struct task_struct *")
+        else:
+            owner = cast("struct task_struct *", rwsem.owner.counter)
+            return owner
+
+
+def get_rwsem_waiter_type(rwsem_waiter: Object) -> str:
+    """
+    Find type of an rwsem waiter
+
+    :param rwsem_waiter: ``struct rwsem_waiter``
+    :returns: str indicating type of rwsem waiter
+    """
+
+    prog = rwsem_waiter.prog_
+    if rwsem_waiter.type.value_() == prog["RWSEM_WAITING_FOR_WRITE"].value_():
+        waiter_type = "down_write"
+    elif rwsem_waiter.type.value_() == prog["RWSEM_WAITING_FOR_READ"].value_():
+        waiter_type = "down_read"
+    else:
+        waiter_type = "waiter type unknown"
+
+    return waiter_type
+
+
+def get_rwsem_waiters_info(rwsem: Object, callstack: int = 0) -> None:
+    """
+    Get a summary of rwsem waiters.
+    The summary consists of ``struct task_struct *``, pid and type of waiters
+
+    :param rwsem: ``struct rw_semaphore *``
+    """
+
+    waiter_type = "none"
+    print("The waiters of rwsem are as follows: ")
+    for waiter in for_each_rwsem_waiter_entity(rwsem):
+        waiter_type = get_rwsem_waiter_type(waiter)
+        task = waiter.task
+        print(
+            f"  ({task.type_.type_name()})0x{task.value_():x}  (pid){task.pid.value_():<8}  (waiter_type){waiter_type:<16}  (cpu){task_cpu(task):<4}  (state){task_state_to_char(task):<4} (wait_time){timestamp_str(task_lastrun2now(task))}"
+        )
```

Comment:
> `({task.type_.type_name()})0x{task.value_():x}`

It looks like you're formatting something like `(struct task_struct *)0xffffdeadbeef`. You can do that a bit easier with:
```
>>> task.format_(dereference=False)
'(struct task_struct *)0xffff897a60da9d00'
```

That being said, that is a very long line of output, and it looks like you're printing a table, just... without the column headers, and you're using the parentheses to mark each field. Maybe this is just a difference of opinion, but I think it's a bit cleaner to just represent things as a table.

```
TASK    PID      TYPE   CPU  ST   WAIT
0xfoo     1 down_read     1   R    20s
```

With the table helpers, we can even handle the case where you want to to a backtrace in between each row, as there are two implementations: `drgn_tools.table.Table`, and `drgn_tools.table.FixedTable`. The latter prints each row as it comes in (with fixed column widths), while the former buffers them all and prints them at the end, when it can determine appropriate column widths for all rows. So you could do something like:

```python
# need FixedTable so each row is printed immediately
tbl = FixedTable(["TASK:>x", "PID:>", "TYPE:16s", "CPU:>", "ST", "WAIT:>"])
for waiter in for_each_rwsem_waiter_entity(rwsem):
    waiter_type = get_rwsem_waiter_type(waiter)
    task = waiter.task
    tbl.row(
        task.value_(),
        task.pid.value_(),
        waiter_type,
        task_cpu(task),
        task_state_to_char(task),
        timestamp_str(task_lastrun2now(task)),
    )
    if callstack:
        bt(task)
tbl.write()  # for compatibility with regular Table API
```

Reply:
I had totally missed the fact that we have these nice table helpers now in drgn_tools. IIRC these came in when Richard was working on dentry helpers. Using table is definitely much better option here and I will integrate your suggestion.